optimize/utils.q

```diff
@@ -0,0 +1,674 @@
+// @param findMin {number[]} The currently calculated minimum values
+// @param highLow {dictionary} Upper and lower bounds of the search space
+// @param cubicCheck {float} Interpolation check parameter
+// @param zoomDict {dictionary} Parameters to be updated as 'zoom' procedure is 
+//   applied to find the optimal value of alpha
+// @returns {boolean} Indication as to if the value of findMin needs to be
+//   pdated 
+i.quadCriteria:{[findMin;highLow;cubicCheck;zoomDict]
+  // On first iteration the initial minimum has not been calculated
+  // as such criteria should exit early to complete the quadratic calculation
+  if[findMin~();:1b];
+  check1:0=zoomDict`idx;
+  check2:findMin>highLow[`low] -cubicCheck;
+  check3:findMin<highLow[`high]+cubicCheck;
+  check1 or check2 or check3
+  }
+
+// @private
+// @kind function
+// @category optimizationUtility
+// @desc Check if the zoom conditions are sufficient
+// @param phi0 {float} Objective function evaluation at index 0
+// @param derPhi0 {float} Derivative of objective function evaluated at index 0
+// @param phiMin {float} 0bjective function evaluated at the current minimum
+// @param findMin {float} The currently calculated minimum value
+// @param zoomDict {dictionary} Parameters to be updated as 'zoom' procedure is
+//   applied to find the optimal value of alpha
+// @param params {dictionary} Parameter dictionary containing the updated/
+//   default information used to modify the behaviour of the system as a whole
+// @returns {boolean} Indication as to if further zooming is required
+i.zoomCriteria1:{[phi0;derPhi0;phiMin;findMin;zoomDict;params]
+  calc:phi0+findMin*derPhi0*params`c1;
+  check1:phiMin>calc;
+  check2:phiMin>=zoomDict`phiLo;
+  check1 or check2
+  }
+
+// @private
+// @kind function
+// @category optimizationUtility
+// @desc Check if the zoom conditions are sufficient
+// @param derPhi0 {float} Derivative of the objective function evaluated at
+//   index 0
+// @param derPhiMin {float} Derivative of the objective function evaluated at
+//   the current minimum
+// @param params {dictionary} Parameter dictionary containing the 
+//   updated/default information used to modify the behaviour of the system 
+//   as a whole
+// @returns {boolean} Indication as to if further zooming is required
+i.zoomCriteria2:{[derPhi0;derPhiMin;params]
+  abs[derPhiMin`derval]<=neg derPhi0*params`c2
+  }
+
+// @private
+// @kind function
+// @category optimizationUtility
+// @desc Check if the zoom conditions are sufficient
+// @param derPhiMin {float} Derivative of the objective function evaluated at 
+//   the current minimum
+// @param alphaDiff {float} Difference between the upper and lower bound of the 
+//   zoom bracket
+// @returns {boolean} Indication as to if further zooming is required
+i.zoomCriteria3:{[derPhiMin;alphaDiff]
+  0<=derPhiMin[`derval]*alphaDiff
+  }
+	   
+// Zoom dictionary 
+
+// @private
+// @kind dictionary
```

Comment:
This isn't of kind dictionary it should be `symbol[]` I think. In line with conversations around the standard I think you should also add the following before the definition of `i.zoomKeys`
```
// @type {symbol[]}
```
The same holds for the remaining items in this file